### DISCUSSION TITLE
Does it still make sense to refetch queries on "focus" by default?

### DISCUSSION THREAD

**Discussion by ilyasmez** [TanStack/query#4797] [Ideas]
Hello 👋 

Right now, React Query has `refetchOnWindowFocus` set to `true` by default. And the default "focus" detection logic looks like this:
https://github.com/TanStack/query/blob/52854795b62946b4a4b73e4bddc6954db77e2058/src/core/focusManager.ts#L58-L74

As you can see, the focus manager listens to `visiblitychange` and `focus` events by default.

💡 **I'm wondering if it makes sense to stop listening to `focus` event and listen only to `visibilitychange`.** 💡

To me, it seems that we're supporting both events just for historical reasons, as `visiblitychange` wasn't fully supported in some old browsers which are dropped now (e.g. IE11 💀).

The `focus` event has many pitfalls, as it's triggered in the following cases:
- Navigating between DevTools and the app.
- When closing an `alert` or `confirm` dialog.
- File upload dialogs.
- When interacting with an iframe.

And this can result in a bad UX.

I know that it's possible to customize this event as shown here: https://tanstack.com/query/v4/docs/react/guides/window-focus-refetching#custom-window-focus-event
But this discussion is about the default behavior. I think the library should provide the best possible experience out of the box.

**Comment by TkDodo** [comment]
> To me, it seems that we're supporting both events just for historical reasons, as visiblitychange wasn't fully supported in some old browsers which are dropped now (e.g. IE11 💀).

That's a very good point I actually wasn't aware of. We're updating our browserlist for v5, so this seems like a good moment to reflect on if that is still needed.

I'm aware of the "`alert` issue" - I didn't know that the `focus` event was the culprit.

What you are saying makes sense to me - would you like to contribute this change to the `v5` branch?

**Reply by TkDodo** [reply]
I guess this would also help with the input type file problem?
- https://github.com/TanStack/query/issues/2960

**Reply by kyawthura-gg** [reply]
@TkDodo I would like to contribute. can you please give me hint or assign me to that issue? Thank you 🙏 .

**Reply by ilyasmez** [reply]
> I guess this would also help with the input type file problem?
> 
> * [Input type=file dialogue triggers windows focus refetch #2960](https://github.com/TanStack/query/issues/2960)

Yes it would, on Chrome for MacOS at least. I'm not sure if there are some browsers that change the visibility state of the document if the file picker is open.

@kyawthura-gg I already have a fix for this. I'll open a PR today and request your review as well if you would like to (co)contribute.

**Reply by kyawthura-gg** [reply]
Yes, sure.

**Reply by ilyasmez** [reply]
I've finally had the time to work on this fix. The PR is here: https://github.com/TanStack/query/pull/4805
